### ISSUE TITLE
How to use lm loss for voken during training?

### ISSUE THREAD

**Issue by TiaoziLiao** [eric-ai-lab/MiniGPT-5#14]
As stated in the paper "During the training, we append the vokens to the positions of ground truth images and train the model to predict vokens within text generation"

So in the interleaved image and text training data, should each image be replaced with the form [img1], [img2], ... [img8]?

For a sample of the following form:
[s][text1] [text2] [text3] [img1], [img2], ... [img8]

The label for calculating L_text is in the following form?
[text2] [text3] [img1], [img2], ... [img8][/s]

will there be two problems:
All images are represented by Vokens arranged in the same way.
Force large models to be generated sequentially [img]

**Comment by KzZheng** [comment]
I would like to clarify that every generated image will be replaced by "[IMG0][IMG1]...[IMG7]" voken sequences. "[IMG0][IMG1]...[IMG7]"  does not mean eight different images. They are a feature sequence used for one image.

Therefore, to answer your questions:
1. The vokens' input embeddings are the same for all images, but the output hidden states depend on the previous context features since LLM is an autoregressive model.
2. [IMG i] vokens need to be generated sequentially since all features of "[IMG0][IMG1]...[IMG7]" are used for a single image. Actually, during the inference, we will force the model to generate "[IMG1]...[IMG7]" when "[IMG0]" is generated. Check the code after [this line](https://github.com/eric-ai-lab/MiniGPT-5/blob/c38b923de830d19581fcdc9cf56aa2453b88f62b/minigpt4/models/mini_gpt5.py#L324) for more details

**Comment by TiaoziLiao** [comment]
Thank you for your answer, I can get what you mean.

I may have misunderstood some of the contents of the paper before. I thought you directly used vokens ([img0] ... [img7]) to generate images. In fact, this paper uses the LLM's output hidden state (with size [seq_len, hidden_size]) corresponding to vokens to generate images.

**Comment by KzZheng** [comment]
Yes, the understanding is correct. Thanks!
